vulkan: add peak performance tuning for ARM Mali GPUs (G720)

Gong-Mi opened the pull request:
### Description
This PR implements specialized performance tuning for ARM Mali G720 GPUs (vendor ID 0x13B5) in the Vulkan backend.

### Changes
- Implement specialized warptile configurations for Mali G720 to optimize compute throughput.
- Force FP32 path by disabling FP16/BF16 based on experimental findings for peak performance on this architecture.
- Limit suballocation block size to 256MB to improve memory stability on mobile devices within the Termux environment.

### Performance Findings (Termux Native Build)
- **1B Models (e.g., Llama 3.2 1B)**: Show significant performance advantages when fully offloaded to the GPU.
- **4B/8B Models**: Require partial CPU offloading to manage memory constraints effectively on typical mobile RAM configurations.
- **Benchmark Command**: `llama-bench -m models/llama-3.2-1b.gguf -p 512 -n 128 -t 4`

These optimizations aim to provide a more usable experience for Android users running LLMs locally via Vulkan on modern Dimensity/Mali-based SoCs.

jeffbolznv commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -2896,7 +2896,16 @@ static void ggml_vk_load_shaders(vk_device& device) {
+            m_warptile_id = m_warptile_mmqid = { 64, 64, 64, 16, 16, 32, 2, 2, 2, 1, 16 };
+            device->suballocation_block_size = 256 * 1024 * 1024;
+            device->fp16 = false; // Force disable FP16
+            device->bf16 = false; // Force disable BF16
```
I'm surprised it's necessary or helpful to force bf16 off.

Gong-Mi commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -2896,7 +2896,16 @@ static void ggml_vk_load_shaders(vk_device& device) {
+            m_warptile_mmq = m_warptile_mmq_int = { 64, 64, 64, 16, 16, 32, 2, 2, 2, 1, 16 };
+            m_warptile = { 64, 64, 64, 16, 16, 32, 2, 2, 2, 1, 16 };
+            m_warptile_id = m_warptile_mmqid = { 64, 64, 64, 16, 16, 32, 2, 2, 2, 1, 16 };
+            device->suballocation_block_size = 256 * 1024 * 1024;
```
build_vulkan/bin/llama-bench -m models/llama-3.2-1b.gguf -p 512 -n 128 -t 3 -ngl 99
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = Mali-G720-Immortalis MC12 (Mali-G720-Immortalis MC12) | uma: 1 | fp16: 1 | bf16: 0 | warp size: 16 | shared memory: 32768 | int dot: 1 | matrix cores: none
| model                          |       size |     params | backend    | ngl | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | --------------: | -------------------: |
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
| llama 1B Q4_K - Medium         | 762.81 MiB |     1.24 B | Vulkan     |  99 |       3 |           pp512 |        151.01 ± 6.52 |
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
ggml_vulkan: Mali G720 Tuning - FORCING FP32 PATH.
| llama 1B Q4_K - Medium         | 762.81 MiB |     1.24 B | Vulkan     |  99 |       3 |           tg128 |         38.83 ± 0.33 |

build: 21844aac3 (7580)
I'm not sure about this performance.

Gong-Mi commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -2896,7 +2896,16 @@ static void ggml_vk_load_shaders(vk_device& device) {
 
         // chip specific tuning
-        if ((device->architecture == AMD_GCN) && (device->driver_id != vk::DriverId::eAmdProprietary)) {
+        if (device->vendor_id == 0x13B5) { // ARM Mali
```
fulfil

jeffbolznv commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -4455,6 +4460,22 @@ static vk_device ggml_vk_get_device(size_t idx) {
+        device->uma = device->properties.deviceType == vk::PhysicalDeviceType::eIntegratedGpu;
+
+        if (device->vendor_id == VK_VENDOR_ID_ARM) {
+            device->suballocation_block_size = 256 * 1024 * 1024;
```
Please move this where the rest of the suballocation_block_size logic is, I don't think it even has any effect here.

Gong-Mi commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -4455,6 +4460,22 @@ static vk_device ggml_vk_get_device(size_t idx) {
+            bfloat16_support = false;
+
+            // Mali is always UMA
+            device->uma = true;
```
okay

jeffbolznv commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -4455,6 +4460,22 @@ static vk_device ggml_vk_get_device(size_t idx) {
+
+            fp16_storage = false;
+            fp16_compute = false;
+            bfloat16_support = false;
```
I'd still like to better understand what these are accomplishing. Which shaders are slower with fp16? You can use GGML_VK_PERF_LOGGER to help identify them.

Gong-Mi replied:
OK

Gong-Mi replied:
Has been merged into one submission, please review.

jeffbolznv commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -4487,6 +4503,11 @@ static vk_device ggml_vk_get_device(size_t idx) {
         device->driver_id = driver_props.driverID;
 
+        if (device->vendor_id == VK_VENDOR_ID_ARM) {
+            // Previously forced FP32 here due to poor FP16 performance on some ARM GPUs.
```
This comment doesn't need to be in the code. Otherwise the change looks fine to me, but @0cc4m should review.

0cc4m reviewed (commented):
Please rebase to resolve the conflict.

0cc4m commented:
Please stop making large changes, it is impossible to review a moving target. Also, make sure to follow https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md, your recent responses were definitely AI and your code changes may also be. This does not give me confidence that the PR actually does what it claims to do.

Gong-Mi committed:
vulkan : identify Mali GPU and enable Android platform headers

Gong-Mi committed:
vulkan : add VK_ARM_data_graph support and fix fa_mask_opt build error

Gong-Mi committed:
vulkan: add dynamic AHB detection and conditional zero-copy logic

- Add CMake check for android/hardware_buffer.h
- Implement runtime detection for integrated GPUs lacking standard zero-copy support
- Enable AHB zero-copy only when standard zero-copy is unavailable on Android

Gong-Mi committed:
vulkan: use hardcoded extension string for AHB to fix cross-platform build

Gong-Mi committed:
vulkan: fix build error by replacing undefined VK_LOG_INFO with std::cerr

Gong-Mi committed:
vulkan: fix AHB header inclusion and add default member initialization

Gong-Mi committed:
vulkan: implement triple-check logic for zero-copy and report [X] if all fail

Gong-Mi committed:
vulkan: protect ARM-specific data graph extensions with macros to fix CI build

BlindDeveloper commented:
@Gong-Mi  Any updates?

Gong-Mi commented:
Superseded: this draft is being abandoned and replaced by a clean PR based on current master.
